### PR TITLE
feat: Microsoft authentication

### LINKED ISSUE #755: Microsoft Teams Authentication
My org is looking to move off of Slack and onto other platforms, and from Google to Microsoft. 

So, while I think this solution is super slick, I'm curious how much trouble it would be--if we did the modifications in a fork--to remove those dependencies (and possibly swap in a Teams integration). 

We'd like to host internally, so auth is less of a concern for us, FWIW.

Thanks in advance.

### PR DESCRIPTION
**Pull request by tommoor**
[>> Setup Guide <<](https://wiki.generaloutline.com/share/dfa77e56-d4d2-4b51-8ff8-84ea6608faa4)

If anyone wants to follow the guide and try this branch out on their own instance it would be much appreciated, I only have one Microsoft team to test with and it would be good to get some broader coverage before merging

closes #755

### CODE REVIEW

File: server/auth/azure.js
Comment (tommoor) [inline server/auth/azure.js:21]: "Microsoft", "Azure", "Teams", "Office 365" – I'm tempted to just do "Microsoft", it seems to cover all the possible options.

File: server/auth/azure.js
Comment (tommoor) [inline server/auth/azure.js:44]: Ideally we'd grab the organization name with an API request, haven't quite figured that out yet. The accessToken returned from oauth isn't working with Microsoft' graph API

Comment (ulentini) [reply server/auth/azure.js:44]: @tommoor the callback signature seems to be wrong.
I tried to implement the same strategy, but as you can see on the strategy page (http://www.passportjs.org/packages/passport-azure-oauth2/), there's no `req` object in the callback, the first parameter is the accessToken. So I guess you're trying to query the graph API with the refresh token

Comment (tommoor) [reply server/auth/azure.js:44]: The startegy callbacks are standardized – `passReqToCallback: true` in the strategy options makes the first parameter always the request.

Comment (ulentini) [reply server/auth/azure.js:44]: Yes, you're right. The problem actually is that that package use the old Azure AD Graph API, instead of the new Microsoft Graph API

Comment (tommoor) [reply server/auth/azure.js:44]: Could you point to that? I didn't think it was using either graph api.

I'll probably fork the whole thing, it's also point to an old login url.

Comment (ulentini) [reply server/auth/azure.js:44]: Sure, here:
https://github.com/AndrewKeig/passport-azure-oauth2/blob/master/lib/index.js#L9
The base URL for oauth requests is `login.windows.net` instead of `graph.microsoft.com`.

File: server/auth/azure.js
Comment (ulentini) [inline server/auth/azure.js:42]: Keep in mind that `profile.email` is not always present. In my company profile I can only see `profile.upn`, which contains my email address. In my personal account (which is a mix of password and github authentication) I have a mess.
I guess a request to `https://graph.microsoft.com/v1.0/me` would be better, since all data seem to be normalized, there

**Comment by tommoor** [comment]
Waiting on Microsoft Partner approval before moving this forward

**Comment by bglidwell** [comment]
Will be able to purchase subscription once this is merged.

**Review by bglidwell — APPROVED** [review]
Successfully authenticated for me

**Comment by bglidwell** [comment]
Any idea when this might be merged in?

**Comment by ulentini** [comment]
Hi, I have 2 issues with this branch (but maybe it's just me)
1. If I try to setup the azure oauth with one of my company azure apps, I cannot login since I get the "Invalid audience" error. This could depend on the different endpoints used to perform the initial authentication (from the passport strategy) VS the following requests to MS Graph api
2. Even after I fixed the first problem, only the first login after the server starts has been successful. After that it fails with a "state mismatch" error. I was able to fix this changing [this line](https://github.com/auth0/passport-azure-ad-oauth2/blob/master/lib/passport-azure-ad-oauth2/strategy.js#L116) in the passport azure package. It's like there's some sort of closure on the `options` object and once the first user generate the state cookie, it gets used for all following users. So, basically, changing that return statement to `return { ...options }` solved the issue for me.

**Comment by tommoor** [comment]
RE 2, weird but it could be the case – other packages do produce a new object, eg:

https://github.com/jaredhanson/passport-google-oauth2/blob/2399ef47f00b47ac587f056f2f100a2c4db81928/lib/strategy.js#L134

I'll fork the dependency it's clearly not maintained and a poor thing to rely on for such a key aspect of the application

**Comment by tommoor** [comment]
@ulentini moved to a fork, and included your fix for the state and the updated authentication domain. I checked that the state is not retained between sign-ins, so looking good now.

**Comment by tommoor** [comment]
@bglidwell this is now deployed to the cloud hosted version of Outline

**Comment by fooness** [comment]
That’s a good step forward!

When might we get authentication methods which are not externalized to Google, Microsoft, or Slack?

Local accounts with Postgres? LDAP? SAML?

https://github.com/outline/outline/pull/1183

https://github.com/outline/outline/issues/1881

**Comment by tommoor** [comment]
LDAP and SAML are included in the enterprise edition and will not be included in the open source codebase – you (or anyone in the future that reads this) can contact hello@getoutline.com to arrange an enterprise license key.

I have detailed what is required for local accounts in the ticket you linked, if the community wants to build that feature now that the authentication providers are pluggable they are welcome to do so!

**Pull request merged**
